### PR TITLE
Drop network transitions on pipe  overflow

### LINKED ISSUE #4119: [Crash] Pipe_lib.Strict_pipe.Overflow
```
[2019-12-25 05:57:35.884978]Init__Coda_run: Unhandled top-level exception: $exn
Generating crash report
$exn = "(monitor.ml.Error\n  (\"Pipe_lib.Strict_pipe.Overflow(\\\"transition frontier controller pipe\\\")\")\n  (\"Raised at file \\\"lib/pipe_lib/strict_pipe.ml\\\", line 211, characters 
8-53\"\n    \"Called from file \\\"src/pipe.ml\\\", line 873, characters 10-13\"\n    \"Called from file \\\"src/job_queue.ml\\\" (inlined), line 131, characters 2-5\"\n    \"Called from fi
le \\\"src/job_queue.ml\\\", line 171, characters 6-47\"\n    \"Caught by monitor coda\"))"
```
```
$ cat coda.version
bd5715a3f00754869498083b2506db776a7047e2
```
[coda_crash_report_2019-12-25_05-57-36.340262.tar.gz](https://github.com/CodaProtocol/coda/files/4000016/coda_crash_report_2019-12-25_05-57-36.340262.tar.gz)

### PR DESCRIPTION
**Pull request by deepthiskumar**
Here's a rough diagram of the pipe connections for network transitions in transition_router, transition_frontier_controller, bootstrap_controller:

output

Based on the flow in this diagram, if a node is catching up then `primary_transitions` pipe will not be processed until `catchup_breadcrumb` pipe is empty.  I think we should instead just drop on overflow and not increase the pipe capacity. If the node is in catchup then the network transitions that are processed in between catchup jobs would probably also trigger catchup. So dropping older transitions may reduce the number of catchup jobs eventually.
Also, change all the pipes in line to drop on overflow and make their capacities equal.

Similarly for bootstrap controller pipe, dropping on overflow will reduce the number of sync ledger queries. In fact, should we reduce the capacity there? Worst case why would we want to check 50 transitions if they are worth getting the root for? 

To summarize, changing all the pipes shown in the diagram to have a capacity of 50 and drop on overflow

This should fix #4119, #8502, #4079, #8738. Could be the cause of #8985
Checklist:

- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them:

### CODE REVIEW

File: src/lib/transition_router/transition_router.ml [hunk 1]
```diff
@@ -13,8 +13,8 @@ type Structured_log_events.t += Starting_transition_frontier_controller
 type Structured_log_events.t += Starting_bootstrap_controller
   [@@deriving register_event {msg= "Starting bootstrap controller phase"}]
 
-let create_bufferred_pipe ?name () =
-  Strict_pipe.create ?name (Buffered (`Capacity 50, `Overflow Crash))
+let create_bufferred_pipe ?(capacity = 50) ?name () =
```
Comment (mrmr1993) [inline src/lib/transition_router/transition_router.ml:16]: Do we need to worry here about saturating the bootstrap/transition frontier controller pipes, and then overflowing the downstream pipes? It seems as though we might now allow more blocks to enter the system than the later pipes have capacity to handle.

In particular, could you generate 100 distinct blocks by re-ordering transactions within a single valid block's diff, and use that to trigger this crash? Perhaps it would be safer here to `Overflow Drop_head`?

Comment (deepthiskumar) [reply src/lib/transition_router/transition_router.ml]: made a flow diagram of pipes and i think it makes sense just to drop head on overflow. Updated the description of the PR accordingly

**Review by mrmr1993 — APPROVED** [review]
Looks good!

**Review by nholland94 — COMMENTED** [review]
I think that most of these pipes which are currently set to `Drop_head` should be set to `Call` with a function that logs when we are dropping information from this pipe. This will help us debug issues related to dropping messages from these pipes in the future.

File: src/lib/transition_frontier_controller/transition_frontier_controller.ml [hunk 2]
```diff
@@ -6,17 +6,18 @@ open O1trace
...
+      (Buffered (`Capacity valid_transition_pipe_capacity, `Overflow Drop_head))
   in
   let primary_transition_pipe_capacity =
     valid_transition_pipe_capacity + List.length collected_transitions
   in
+  (* Ok to drop on overflow- catchup will be triggered if required*)
   let primary_transition_reader, primary_transition_writer =
     Strict_pipe.create ~name:"primary transitions"
-      (Buffered (`Capacity primary_transition_pipe_capacity, `Overflow Crash))
+      (Buffered (`Capacity primary_transition_pipe_capacity, `Overflow Drop_head))
```
Comment (nholland94) [inline src/lib/transition_frontier_controller/transition_frontier_controller.ml:20]: When these transitions are dropped from the pipes, we should probably fill the validation callback for libp2p's sake and tell it that to ignore that message. This will make the logs more useful since we will filter out any validation callback timeouts caused by dropping these transitions. You can provide a function to call when messages are dropped from the pipe by using `Overflow (Call f)`.

**Review by nholland94 — CHANGES_REQUESTED** [review]
Still thinking through the consequences of this, and I think it's generally fine. Left some notes on better handling dropped messages from these pipes.

File: src/lib/transition_frontier_controller/transition_frontier_controller.ml (hunk 2 above)
Comment (deepthiskumar) [reply src/lib/transition_frontier_controller/transition_frontier_controller.ml:20]: There are warning logs with the `Drop_head option.  `Call f` doesn't drop but just executes f on the data it is trying to write. Are you suggesting we drop the latest message instead of oldest message that gets removed with the `Drop_head` variant?

**Review by nholland94 — APPROVED** [review]
